libs/langchain/langchain/agents/output_parsers/openai_functions.py

```diff
@@ -0,0 +1,89 @@
+import json
+from dataclasses import dataclass
+from json import JSONDecodeError
+from typing import List, Union
+
+from langchain.agents.agent import AgentOutputParser
+from langchain.schema import (
+    AgentAction,
+    AgentFinish,
+    OutputParserException,
+)
+from langchain.schema.messages import (
+    AIMessage,
+    BaseMessage,
+)
+from langchain.schema.output import ChatGeneration, Generation
+
+
+@dataclass
+class _FunctionsAgentAction(AgentAction):
+    message_log: List[BaseMessage]
+
+
+class OpenAIFunctionsAgentOutputParser(AgentOutputParser):
+    """Parses a message into agent action/finish.
+
+    Is meant to be used with OpenAI models, as it relies on the specific
+    function_call parameter from OpenAI to convey what tools to use.
+
+    If a function_call parameter is passed, then that is used to get
+    the tool and tool input.
+
+    If one is not passed, then the AIMessage is assumed to be the final output.
+    """
+
+    @property
+    def _type(self) -> str:
+        return "openai-functions-agent"
+
+    @staticmethod
+    def _parse_ai_message(message: BaseMessage) -> Union[AgentAction, AgentFinish]:
```

Comment:
what's the advantage v being a standalone function?

Reply:
could be, main thing was just making sure to use it here and the openai functions agent. could make public function not on class. no preference really